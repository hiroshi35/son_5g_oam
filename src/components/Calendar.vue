<template>
  <div class="sc-calendar">
    <div v-if="externalEvents.length || addEvents" ref="externals">
      <div class="row">
        <div class="col-md-2 sc-events">
          <h4>External Events</h4>
          <div class='fc-event' v-for="event in externalEvents" :key="event.id" :data-id="event.id"
               :data-color="event.color ? event.color : '#2e99b3'"
               :style="{backgroundColor: event.color ? event.color : '#2e99b3',borderColor: event.color ? event.color : '#2e99b3'}">
            {{event.title}}
          </div>
          <div v-show="externalEvents.length" class="divider"></div>
          <button v-show="addEvents" type="button" class="btn btn-red right-angular btn-block margin-bottom-15"
                  @click="addEvent = !addEvent">{{addEvent ? 'CANCEL ' : 'ADD EVENT '}}
            <i aria-hidden="true" class="fa" :class="addEvent ? 'fa-times' : 'fa-plus'">
          </i></button>
          <div v-show="addEvent" class="new-event">
            <h4>Add New Event</h4>
            <form action="#">
              <div v-if="!idAutoGeneration" class="form-group" :class="{'has-error' : eventIdError}"><input
                class="form-control" placeholder="#ID" v-model="eventId">
                <p v-if="eventIdError" class="help-block">ID is required</p>
              </div>
              <div class="form-group" :class="{'has-error' : eventNameError}">
                <input class="form-control" placeholder="Enter Event Name" v-model="eventName">
                <p v-if="eventNameError" class="help-block">Event Name is required</p>
              </div>
              <div class="form-group" :class="{'has-error' : eventColorError}">
                <colorpicker :icon="true" value="#2e99b3" v-model="eventColor"></colorpicker>
                <p v-if="eventColorError" class="help-block">Invalid Color</p>
              </div>
              <button type="button" class="btn btn-teal right-angular btn-block push-event-button btn-block margin-bottom-15" @click="pushEvent">ADD EVENT</button>
            </form>
          </div>
        </div>
        <div class="col-md-10">
          <div ref="calendar"></div>
        </div>
      </div>
    </div>
    <div v-if="!externalEvents.length && !addEvents" ref="calendar"></div>
  </div>
</template>


<script>
  import Colorpicker from './Colorpicker.vue'
  import 'fullcalendar'
  import 'moment'
  import 'jquery-ui/ui/widgets/draggable'

  export default {
    props: {
      options: {
        type: Object,
        default () {
          return {
            header: {
              left: 'prev,next today',
              center: 'title',
              right: 'month,agendaWeek,agendaDay'
            }
          }
        }
      },
      events: {
        type: Array
      },
      draggable: {
        type: Boolean,
        default: false
      },
      externalEvents: {
        type: Array,
        default: () => []

      },
      aspectRatio: {
        type: Number,
        default: 1.35
      },
      idAutoGeneration: {
        type: Boolean,
        default: true
      },
      deleteEvents: {
        type: Boolean,
        default: false
      },
      addEvents: {
        type: Boolean,
        default: false
      },
      onDrop: {
        type: Function,
        default: null
      },
      onResize: {
        type: Function
      },
      onDropExternal: {
        type: Function
      }
    },
    computed: {
      optionObj () {
        let obj = this.options
        let vm = this

        obj.aspectRatio = this.aspectRatio
        if (this.events) {
          obj.events = this.events
        }
        if (this.draggable) {
          obj.editable = true
          obj.droppable = true
          obj.eventDrop = function (event, dayDelta, minuteDelta, allDay, revertFunc) {
            vm.events.forEach(function (element, index, array) {
              if (event.id === element.id) {
                element.start = event.start.format()
                if (event.end) {
                  element.end = event.end.format()
                } else {
                  delete element.end
                }
                if (typeof vm.onDrop === 'function') {
                  vm.onDrop(element);
                }
              }
            })
          }
          obj.eventResize = function (event, delta, revertFunc) {
            vm.events.forEach(function (element, index, array) {
              if (event.id === element.id) {
                element.end = event.end.format()
                if (typeof vm.onResize === 'function') {
                  vm.onResize(element)
                }
              }
            })
          }

          obj.drop = function (date) {
            $(this).remove()
            let id = $(this).data('id')
            vm.$emit('eventAdded', {
              id: id,
              title: $.trim($(this).text()),
              start: date.format(),
              color: $(this).data('color')
            })
            let index = vm.getExternalEventIndexById(id)
            vm.externalEvents[index].inEvents = true
            if (typeof vm.onDropExternal === 'function') {
              vm.onDropExternal(vm.externalEvents[index])
            }
          }
        }
        obj.eventRender = function (event, element) {
          if (vm.deleteEvents) {
            element.find('.fc-bg').css('pointer-events', 'none')
            element.append('<div style=\'position:absolute;bottom:0px;right:0px;z-index: 9999;\'><button type=\'button\' class=\'btn btn-block btnDeleteEvent\'><i class=\'fa fa-times\' aria-hidden=\'true\'></i></button></div>')
            element.find('.btnDeleteEvent').click(function (e) {
              e.preventDefault()
              $(vm.$refs.calendar).fullCalendar('removeEvents', event._id)
              vm.events.remove(vm.getEventIndexById(event.id))
            })
          }
        }
        return obj
      }
    },
    mounted () {
      $(this.$refs.calendar).fullCalendar(this.optionObj)
      if (this.externalEvents) {
        this.initExternalEvents()
      }
    },
    updated () {
      this.initExternalEvents()
    },
    data () {
      return {
        addEvent: false,
        eventName: '',
        eventId: '',
        eventColor: '#2e99b3',
        eventNameError: false,
        eventColorError: false,
        eventIdError: false
      }
    },
    methods: {
      pushEvent () {
        let hasError = false
        this.eventIdError = false
        this.eventNameError = false
        this.eventColorError = false
        if (!this.idAutoGeneration && !this.eventId) {
          this.eventIdError = true
          hasError = true
        }
        if (!this.eventName) {
          this.eventNameError = true
          hasError = true
        }
        if (!this.eventColor) {
          this.eventColorError = true
          hasError = true
        } else {
          let isValidColor = /(^#[0-9A-Fa-f]{6})|(^#[0-9A-Fa-f]{3})/.test(this.eventColor)
          if (!isValidColor) {
            this.eventColorError = true
            hasError = true
          }
        }
        if (hasError) {
          return false
        }
        this.$emit('externalEventAdded', {
          id: this.idAutoGeneration ? 'id' + (new Date()).getTime() : this.eventId,
          title: this.eventName,
          color: this.eventColor
        })
        this.addEvent = false
        this.eventName = ''
        this.eventId = ''
      },
      initExternalEvents () {
        $('.sc-events .fc-event').each(function () {
          $(this).data('event', {
            id: $(this).data('id'),
            title: $.trim($(this).text()),
            stick: true,
            color: $(this).data('color')
          })
          $(this).draggable({
            zIndex: 999,
            revert: true,
            revertDuration: 0
          })
        })
      },
      getEventIndexById (id) {
        let indx = false
        this.events.forEach(function (element, index, array) {
          console.log(id, element.id);
          if (id === element.id) {
            indx = index
          }
        })
        return indx
      },
      getExternalEventIndexById (id) {
        let indx = false
        this.externalEvents.forEach(function (element, index, array) {
          if (id === element.id) {
            indx = index
          }
        })
        return indx
      }
    },
    components: {
      Colorpicker
    }
  }
</script>

<style lang="scss">
  @import "~fullcalendar/dist/fullcalendar";

  .fc-event {
    border-radius: 0px !important;
  }

  .fc-event .fc-content {
    font-size: 12px;
    font-family: Lato, sans-serif;
  }

  .fc-widget-header {
    font-family: Lato, sans-serif;
    font-weight: 300;
  }

  .fc-widget-content {
    font-family: Lato, sans-serif;
    font-weight: 300;
    font-size: 11px;
  }

  .fc-day-grid-event {
    margin: 1px 2px 0;
    padding: 4px 4px !important;
  }

  .fc-today {
    opacity: .2;
    border: none;
    background-color: #78d657 !important;
  }

  .fc-day-top {
    opacity: 1;
    background: none !important;
  }

  .fc-time-grid-event.fc-v-event.fc-event {
    border-radius: 4px;
    border: none;
    padding: 5px;
    opacity: .65;
    left: 5% !important;
    right: 5% !important;
  }

  .fc-event {
    font-size: inherit !important;
  }

  .fc td, .fc th {
    border-style: none !important;
    border-width: 1px !important;
    padding: 0 !important;
    vertical-align: top !important;
  }

  .fc-event .fc-bg {
    z-index: 1 !important;
    background: inherit !important;
    opacity: .25 !important;
  }

  .fc-time-grid-event .fc-time {
    font-weight: normal !important;
  }

  .fc-ltr .fc-h-event.fc-not-end, .fc-rtl .fc-h-event.fc-not-start {
    opacity: .65 !important;
    margin-left: 12px !important;
    padding: 5px ! important;
  }

  .fc-day-grid-event.fc-h-event.fc-event.fc-not-start.fc-end {
    opacity: .65 !important;
    margin-left: 12px !important;
    padding: 5px ! important;
  }

  .sc-events .fc-event {
    cursor: pointer;
    margin-bottom: 5px;
    padding: 2px 4px;

  }

  .fc-center h2 {
    font-family: Lato, sans-serif;
    font-weight: 300;
  }

  .sc-calendar h4 {
    font-family: Lato, sans-serif;
    font-weight: 300;
  }

  .fc-state-default {
    background: none !important;
  }

  .fc-state-down, .fc-state-active {
    background-color: #f7f7f7 !important;
    box-shadow: none !important;
  }

  .sc-events {
    position: relative;
  }

  .btnDeleteEvent {
    background: none;
    opacity: .4;
  }

  .btnDeleteEvent:hover {
    color: #fff;
    opacity: 1;
  }
</style>
