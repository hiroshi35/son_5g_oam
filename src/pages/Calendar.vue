<template>
  <div>
    <div class="container-fluid">
      <!-- Page Heading -->
      <page-header title="Calendar"></page-header>
      <div class="row">
        <div class="col-md-12">
          <panel type="panel-white right-angular">
            <h3 slot="header">
              FullCalendar
            </h3>
            <calendar :events="values.events"
                      :draggable="true"
                      :external-events="values.externalEvents"
                      :delete-events="true"
                      :add-events="true"
                      :on-drop-external="values.onDrop"
                      @eventAdded="addEvent"
                      @externalEventAdded="addExternalEvent"
            ></calendar>
          </panel>
        </div>
      </div>
    </div>
    <!-- /.container-fluid -->
  </div>
</template>

<script>
  import Calendar from '../components/Calendar.vue'

  export default {
    methods: {
      addEvent (e) {
        this.values.events.push(e)
      },
      addExternalEvent (e) {
        this.values.externalEvents.push(e)
      }
    },
    data () {
      return {
        values: {
          events: [
            {
              id: 1,
              title: 'Event',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() - 3),
              color: '#e53935'
            },
            {
              id: 2,
              title: 'Long Event',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() - 1),
              end: new Date(Date.now()).setDate(new Date(Date.now()).getDate() + 2),
              color: '#90caf9'
            },
            {
              id: 199,
              title: 'Lunch with Jeffrey',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() + 4),
              color: '#80cbc4'
            },
            {
              id: 999,
              title: 'Party',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() + 1),
              color: '#ffd54f'
            },
            {
              id: 3,
              title: 'Meeting',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() - 2),
              color: '#66bb6a'
            },
            {
              id: 4,
              title: 'Lunch',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() - 10),
              color: '#ba68c8'
            },
            {
              id: 5,
              title: 'Birthday Party',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() + 6)
            },
            {
              id: 6,
              title: 'Click for Google',
              url: 'http://google.com/',
              start: new Date(Date.now()).setDate(new Date(Date.now()).getDate() + 10)
            }
          ],
          externalEvents: [
            {
              id: 7,
              title: 'External Event 1',
              color: '#fa7252'
            },
            {
              id: 8,
              title: 'External Event 2',
              color: '#dc5e5a'
            },
            {
              id: 9,
              title: 'External Event 3',
              color: '#3ab39b'
            },
            {
              id: 10,
              title: 'External Event 4'
            },
            {
              id: 11,
              title: 'External Event 5',
              color: '#aebd38'
            }
          ],
          onDrop: function (event) {
            console.log(event);
          }
        }
      }
    },
    components: {
      Calendar
    }
  }
</script>