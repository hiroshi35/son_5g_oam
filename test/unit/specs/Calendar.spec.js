import { shallow } from 'vue-test-utils'
import Calendar from '@/components/Calendar.vue'
import {fullcalendar} from 'jquery'

describe('Calendar.vue', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(Calendar, {
      propsData: {
        events: [
          {
            id: 1,
            title: 'Event',
            start: '2018-04-04',
            color: '#e53935'
          }
        ],
        externalEvents: [
          {
            id: 7,
            title: 'External Event 1',
            color: '#fa7252'
          }
        ]
      }
    })
  })
  it('should be vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  test('$().fullCalendar() should be called with correct options', () => {
    expect(fullcalendar).toBeCalledWith(wrapper.vm.optionObj)
  })
  it('has correct pros', () => {
    expect(wrapper.props().events).toEqual([{id: 1, title: 'Event', start: '2018-04-04', color: '#e53935'}])
    expect(wrapper.props().externalEvents).toEqual([{id: 7, title: 'External Event 1', color: '#fa7252'}])
  })
  test('push external event', () => {
    wrapper.find('.push-event-button').trigger('click')
    expect(wrapper.vm.eventNameError).toBeTruthy();
    wrapper.setData({ eventName: 'bar' })
    wrapper.find('.push-event-button').trigger('click')
    expect(wrapper.emitted().externalEventAdded).toBeTruthy()
  })
})
