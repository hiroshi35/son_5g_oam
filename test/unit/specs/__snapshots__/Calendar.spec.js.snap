// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar.vue has the expected html structure 1`] = `
<div
  class="sc-calendar"
>
  <div>
    <div
      class="row"
    >
      <div
        class="col-md-2 sc-events"
      >
        <h4>
          External Events
        </h4>
         
        <div
          class="fc-event"
          data-color="#fa7252"
          data-id="7"
          style="background-color: rgb(250, 114, 82); border-color: #fa7252;"
        >
          
          External Event 1
        
        </div>
         
        <div
          class="divider"
        />
         
        <button
          class="btn btn-red right-angular btn-block margin-bottom-15"
          style="display: none;"
          type="button"
        >
          ADD EVENT 
          
          <i
            aria-hidden="true"
            class="fa fa-plus"
          />
        </button>
         
        <div
          class="new-event"
          style="display: none;"
        >
          <h4>
            Add New Event
          </h4>
           
          <form
            action="#"
          >
            <!---->
             
            <div
              class="form-group"
            >
              <input
                class="form-control"
                placeholder="Enter Event Name"
              />
               
              <!---->
            </div>
             
            <div
              class="form-group"
            >
              <!---->
               
              <!---->
            </div>
             
            <button
              class="btn btn-teal right-angular btn-block push-event-button btn-block margin-bottom-15"
              type="button"
            >
              ADD EVENT
            </button>
          </form>
        </div>
      </div>
       
      <div
        class="col-md-10"
      >
        <div />
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
